ceph_iscsi_config/device_status.py

```diff
@@ -0,0 +1,161 @@
+            return
+
+        counter_dict = self.gw_counter_lookup.get(gw)
+        if counter_dict is None:
+            counter_dict = {}
+
+        for name, val in status.items():
+            if name == "lock_owner" and val == "true":
+                dt = datetime.strptime(status_stamp, "%Y-%m-%dT%H:%M:%S.%f%z")
+                if self.lock_owner_timestamp is None or dt > self.lock_owner_timestamp:
+                    self.lock_owner_timestamp = dt
+                    self.lock_owner = gw
+                    self.stable_cnt = 0
+                continue
+
+            if name not in ["cmd_timed_out_cnt", "conn_lost_cnt", "lock_lost_cnt"]:
+                continue
+
+            stat_cnt = counter_dict.get(name)
+            if stat_cnt is None:
+                stat_cnt = StatusCounter(name, int(val))
+
+            stat_cnt.cnt = int(val)
+            # TODO:
+            # If we detect a degraded state, we can throttle the path here.
+            self.check_for_degraded_state(stat_cnt)
+            stat_cnt.last_cnt = stat_cnt.cnt
+
+            counter_dict[name] = stat_cnt
+
+        self.gw_counter_lookup[gw] = counter_dict
+
+
+class DeviceStatusWatcher(threading.Thread):
+    def __init__(self, logger):
+        threading.Thread.__init__(self)
+        self.logger = logger
+        self.daemon = True
+        self.cluster = None
+        self.status_lookup = {}
+
+    def get_dev_status(self, image_name):
+        return self.status_lookup.get(image_name)
+
+    def exit(self):
+        if self.cluster:
+            self.cluster.shutdown()
+
+    def run(self):
+        self.cluster = rados.Rados(conffile=settings.config.cephconf,
+                                   name=settings.config.cluster_client_name)
+        self.cluster.connect()
+
+        while True:
+            time.sleep(settings.config.status_check_interval)
+
+            cmd = json.dumps({"prefix": "service status", "format": "json"})
+
+            ret, outb, outs = self.cluster.mgr_command(cmd, b'')
+            if ret != 0:
+                self.logger.error("mon command failed {}".format(ret))
+                continue
+
+            svc = json.loads(outb).get('tcmu-runner')
+            if svc is None:
+                self.logger.warning("there is no tcmu-runner data avaliable")
+                self.state = "Unknown"
+                continue
+
+            image_names_dict = {}
+            for daemon, daemon_info in svc.items():
+                gw, image_name = daemon.split(":", 1)
+                image_names_dict[image_name] = image_name
+
+                dev_status = self.get_dev_status(image_name)
+                if dev_status is None:
+                    dev_status = TcmuDevStatusTracker(image_name)
+                    self.status_lookup[image_name] = dev_status
+
+                dev_status.update_status(gw, daemon_info.get('status'), daemon_info.get('status_stamp'))
```

Comment:
Tox: `./ceph_iscsi_config/device_status.py:141:100: E501 line too long (104 > 99 characters)`

Reply:
Sure, I will fix it. Didn't get this warning locally.